python/spark_sklearn/keyed_models.py

```diff
@@ -0,0 +1,407 @@
+        :raise ValueError: if `sklearnEstimator` does not derive from `sklearn.base.BaseEstimator`.
+        :raise ValueError: if `keyCols` is empty.
+        :raise ValueError: if any column has the name `"estimator"` (
+        :raise ValueError: if reflection checks indicate that parameter estimator is not equipped
+        with a `fit()` method or an appropriate transformation/prediction method.
+        """
+        if sklearnEstimator is None:
+            raise ValueError("sklearnEstimator should not be unspecified or None")
+        if not isinstance(sklearnEstimator, sklearn.base.BaseEstimator):
+            raise ValueError("sklearnEstimator should be an sklearn.base.BaseEstimator")
+        if len(keyCols) == 0:
+            raise ValueError("keyCols should not be empty")
+        if "estimator" in keyCols + [xCol, yCol]:
+            raise ValueError("keyCols should not contain a column named \"estimator\"")
+
+        for paramName, paramSpec in KeyedEstimator.paramSpecs.items():
+            setattr(self, paramName, Param(Params._dummy(), paramName, paramSpec["doc"]))
+        super(KeyedEstimator, self).__init__()
+        self._setDefault(**{paramName: paramSpec["default"]
+                            for paramName, paramSpec in KeyedEstimator.paramSpecs.items()
+                            if "default" in paramSpec})
+        kwargs = KeyedEstimator._inferredParams(self.__init__._input_kwargs)
+        self._set(**kwargs)
+
+        self._verifyEstimatorType()
+
+    @staticmethod
+    def _inferredParams(inputParams):
+        if "yCol" in inputParams:
+            inputParams["estimatorType"] = "predictor"
+        return inputParams
+
+    def _verifyEstimatorType(self):
+        estimatorType = self.getOrDefault("estimatorType")
+        estimator = self.getOrDefault("sklearnEstimator")
+
+        if not hasattr(estimator, "fit"):
+            raise ValueError("sklearnEstimator missing fit()")
+
+        if estimatorType == "transformer":
+            if not hasattr(estimator, "transform"):
+                raise ValueError("estimatorType assumed to be transformer, but " +
+                                 "sklearnEstimator is missing transform()")
+        elif estimatorType == "predictor":
+            if not hasattr(estimator, "predict"):
+                raise ValueError("estimatorType assumed to be predictorr, but " +
+                                 "sklearnEstimator is missing predict()")
+        else:
+            raise ValueError("estimatorType {} is unknown".format(estimatorType))
+
+    def _fit(self, dataset):
+        keyCols = self.getOrDefault("keyCols")
+        xCol = self.getOrDefault("xCol")
+        yCol = self.getOrDefault("yCol")
+        isLabelled = yCol is not None
+        assert isLabelled == (self.getOrDefault("estimatorType") == "predictor")
+        _validateXCol(dataset.schema, xCol)
+
+        cols = keyCols[:]
+        cols.append(xCol)
+        if isLabelled: cols.append(yCol)
```

Comment:
I checked the pyspark code base and we do not have such one-liners. New line please.